File: online-judge/4th-semester/exam3/699_TheFallingLeaves.py
from sys import stdin, stdout

MAX_LENGTH = 300

leaves = [0 for _ in range(MAX_LENGTH)]
cin = 0
line = []

def reset():
    for i in range(MAX_LENGTH):
        leaves[i] = 0

def solve(index):
    global cin
    value = line[cin]
    #print(cin)
    cin += 1
    if(value != -1):
        leaves[index] += value
        solve(index - 1)
        solve(index + 1) 

def main():
    global cin, line
    cases = 1
    mid = MAX_LENGTH // 2
    line = list(map(int, stdin.readline().split()))
    while(line[0] != -1):
        #print(line)
        cin = 1
        reset()
        leaves[mid] = line[0]
        solve(mid - 1)
        solve(mid + 1)

        i = 0
        value = leaves[0]
        while(value == 0):
            i += 1
            value = leaves[i]
        
        print(f"Case {cases}:")
        cases += 1

        while(value != 0):
            stdout.write(str(value))
            i += 1
            value = leaves[i]
            if(value != 0):
                stdout.write(" ")
        print("\n")
        line = list(map(int, stdin.readline().split()))

main() 

File: online-judge/4th-semester/hw6/699_TheFallingLeaves.cpp
#include <iostream>

using namespace std;

const int maxLength = 300;

int leaves[maxLength];

void reset()
{
    for(int i = 0; i < maxLength; i++)
    {
        leaves[i] = 0;
    }
}

void solve(int index)
{
    int value;
    cin >> value;
    if(value != -1)
    {
        leaves[index] += value;
        solve(index - 1);
        solve(index + 1); 
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);
    int root, i, value, cases = 1;
    int mid = maxLength / 2;
    cin >> root;
    while(root != -1)
    {
        reset();
        leaves[mid] = root;
        solve(mid - 1);
        solve(mid + 1);

        i = 0;
        value = leaves[0];
        while(value == 0)
        {
            i++;
            value = leaves[i];
        }
        cout << "Case " << cases++ << ":" << endl;
        while(value != 0)
        {
            cout << value;
            i++;
            value = leaves[i];
            if(value != 0)
            {
                cout << " ";
            }
        }
        cout << "\n\n";
        cin >> root;
    }

    return 0;
}